node/pkg/watchers/evm/ccq_backfill.go

```diff
@@ -0,0 +1,316 @@
+package evm
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/certusone/wormhole/node/pkg/common"
+	"github.com/certusone/wormhole/node/pkg/watchers/evm/connectors"
+
+	"go.uber.org/zap"
+
+	ethHexUtil "github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+	ethRpc "github.com/ethereum/go-ethereum/rpc"
+)
+
+const CCQ_MAX_BATCH_SIZE = int64(1000)
+const CCQ_TIMESTAMP_RANGE_IN_SECONDS = uint64(30 * 60)
+
+type (
+	// ccqBackfillRequest represents a request to backfill the cache. It is the payload on the request channel.
+	ccqBackfillRequest struct {
+		timestamp uint64
+	}
+
+	// ccqBatchResult is the result of each query in a batch.
+	ccqBatchResult struct {
+		result ccqBlockMarshaller
+		err    error
+	}
+
+	// ccqBlockMarshaller is used to marshal the query results.
+	ccqBlockMarshaller struct {
+		Number ethHexUtil.Uint64
+		Time   ethHexUtil.Uint64 `json:"timestamp"`
+		// Hash   ethCommon.Hash    `json:"hash"`
+	}
+)
+
+// ccqRequestBackfill submits a request to backfill a gap in the timestamp cache. Note that the timestamp passed in should be in seconds, as expected by the timestamp cache.
+func (w *Watcher) ccqRequestBackfill(timestamp uint64) {
+	select {
+	case w.ccqBackfillChannel <- &ccqBackfillRequest{timestamp: timestamp}:
+		w.ccqLogger.Debug("published backfill request", zap.Uint64("timestamp", timestamp))
+	default:
+		// This will get retried next interval.
+		w.ccqLogger.Error("failed to post backfill request, will get retried next interval", zap.Uint64("timestamp", timestamp))
+	}
+}
+
+// ccqBackfillStart initializes the timestamp cache by backfilling some history and starting a routine to handle backfill requests
+// when a timestamp is not in the cache. This function does not return errors because we don't want to prevent the watcher from
+// coming up if we can't backfill the cache. We just disable backfilling and hope for the best.
+func (w *Watcher) ccqBackfillStart(ctx context.Context, errC chan error) {
+	if err := w.ccqBackfillInit(ctx); err != nil {
+		w.ccqLogger.Error("failed to backfill timestamp cache, disabling backfilling", zap.Error(err))
+		w.ccqBackfillCache = false
+		return
+	}
+
+	common.RunWithScissors(ctx, errC, "ccq_backfiller", func(ctx context.Context) error {
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case evt := <-w.ccqBackfillChannel:
+				w.ccqPerformBackfill(ctx, evt)
+			}
+		}
+	})
+}
+
+// ccqBackfillInit determines the maximum batch size to be used for backfilling the cache. It also loads the initial batch of timestamps.
+func (w *Watcher) ccqBackfillInit(ctx context.Context) error {
+	// Get the latest block so we can use that as the starting point in our cache.
+	latestBlock, err := connectors.GetLatestBlock(ctx, w.ccqLogger, w.ethConn)
+	if err != nil {
+		return fmt.Errorf("failed to look up latest block: %w", err)
+	}
+	latestBlockNum := int64(latestBlock.Number.Uint64())
```

Comment:
is there a reason why we use `int64(latestBlock.Number.Uint64())` instead of `latestBlock.Number.Int64()`?  
if the value is greater than `9223372036854775807` I think both gives us issue so wondering if there's a reason

Reply:
Changed it.